File: isma-intg-lib/isma-intg-lib-rk2/src/main/java/ru/nstu/isma/intg/lib/rungeKutta/rk22/internal/Rk2AccuracyIntgController.java
package ru.nstu.isma.intg.lib.rungeKutta.rk22.internal;

import ru.nstu.isma.intg.core.methods.BaseAccuracyIntgController;

/**
 * @author Mariya Nasyrova
 * @since 20.08.14
 */


public class Rk2AccuracyIntgController extends BaseAccuracyIntgController {
    @Override protected double getLocalErrorEstimation(double[] k) {
        return 0.5 * (k[1] - k[0]);
    }
    @Override public double getActualAccuracy(double step, double localError) {
		return step * localError;
	}
	@Override public double getQ(double accuracy, double actualAccuracy) {
		if (actualAccuracy == 0) {
			return 1.0;
		}
		return Math.sqrt(accuracy / actualAccuracy);
	}
}

File: isma-intg-lib/isma-intg-lib-rk2/src/main/kotlin/ru/nstu/isma/intg/lib/rungeKutta/rk22/internal/Rk2AccuracyIntgController.kt
package ru.nstu.isma.intg.lib.rungeKutta.rk22.internal

import ru.nstu.isma.intg.core.methods.BaseAccuracyIntgController
import kotlin.math.sqrt

/**
 * @author Mariya Nasyrova
 * @since 20.08.14
 */
class Rk2AccuracyIntgController : BaseAccuracyIntgController() {
    override fun getLocalErrorEstimation(deStages: DoubleArray): Double {
        return 0.5 * (deStages[1] - deStages[0])
    }

    public override fun getActualAccuracy(step: Double, localError: Double): Double {
        return step * localError
    }

    public override fun getQ(accuracy: Double, actualAccuracy: Double): Double {
        if (actualAccuracy == 0.0) {
            return 1.0
        }
        return sqrt(accuracy / actualAccuracy)
    }
}